docs: change the remix button from icon to text

minaminao opened the pull request:
Fixes #13948, closes #13951

From the discussion in https://github.com/ethereum/solidity/pull/13951#issuecomment-1508177321 , this PR changes the remix button from an icon to a text. Also, it adds `_blank` so that a new Remix page is opened when the button is pressed.

The style (colors, margins, etc.) may have room for change.

CC: @r0qs

minaminao committed:
docs: change remix button to text button

NunoFilipeSantos commented:
Please trim the trailing whitespaces.

minaminao committed:
docs(fix): trim trailing whitespaces

NunoFilipeSantos commented:
@minaminao have you tested how a PDF export looks like after this change?

minaminao commented:
Yes.

Example:
![image](https://user-images.githubusercontent.com/20497787/232047328-a5b7e7d8-778d-444a-835a-4c5cf71daced.png)

FYI: If the format is PDF, these buttons are not inserted because the following process exists in remix_code_links.py and`app.builder.format` is `latex`.

```py
def insert_remix_link(app, doctree, solidity_version):
    if app.builder.format != 'html' or app.builder.name == 'epub':
        return
```

cameel commented on docs/_static/css/dark.css:
```diff
@@ -626,12 +626,11 @@ code.docutils.literal.notranslate {
-a.remix-link {
-    filter: invert(1); /* The icon is black. In dark mode we want it white. */
+
+a.remix-link,
+a.remix-link:not(:visited) {
+    color: black !important;
 }
```
I think this breaks the styling for visited Remix links. At least in the preview I can't see any difference in appearance between visited and non-visited ones.

minaminao replied:
The styling for visited Remix links is specified by `a.remix-link`, so I think this doesn't break it. The reason why I added `a.remix-link:not(:visited)` is to overwrite the following style:

```css
.rst-content a:not(:visited) {
    color: #aaddff !important;
}
```

In other words, if this code doesn't exist, `a.remix-link:not(:visited)` is not necessary.

minaminao replied:
There may be a better selector.

minaminao replied:
For example:
```
.remix-link-container a.remix-link {
    color: black !important;
}
```

This selector merges `a.remix-link` with `a.remix-link:not(:visited)`.

cameel replied:
> The styling for visited Remix links is specified by `a.remix-link`, so I think this doesn't break it.

It does, at least in Firefox. It overrides color on both `a` and `a:visited`. I think you need to style `a.remix-link:visited` separately.

By the way, forcing it with `!important` isn't really needed, is it? The rule still applies when I remove it.

cameel replied:
Ah, I guess you're using dark mode? Haven't really checked that. It's broken in the normal mode.

And I see that the rule you pointed out has `!important`, which I guess is why you needed it here. But (if possible) it would be better to just remove it from the original rule instead. The problem with `!important` is that it's a quick and easy initially, but once it's on a lot of rules, it's chaos :)

minaminao replied:
~~Do you mean that clicked open-in-remix buttons should have a separate style? This button is represented by `<a>`, but I consider this a button, not a link (This may be contrary to HTML conventions, and maybe we should use `<button>`, but I don't know how to use it in Sphinx), so I think that colors don't need to be separated. Should we separate the colors?~~

>By the way, forcing it with !important isn't really needed, is it? The rule still applies when I remove it.

~~It's `a.remix-link, a.remix-link:not(:visited)` in dark mode, right? In my environment (Chrome), without the `!important`, the text color of the button is light blue. By the way, I don't quite like adding `!important`. I would remove them if I could. (There are `!importants` all over the CSS files before this change.)~~

![image](https://user-images.githubusercontent.com/20497787/232351249-5a761c50-b76c-4c61-b2a3-e0908d323a75.png)

minaminao replied:
~~You may know, but just in case, every time you change the CSS files, `make clean` is required before `make html`.~~

minaminao replied:
Sorry! GitHub didn't show your last reply. Please ignore the two comments above :)

minaminao replied:
I thought you were talking about dark.css. How does it break in the normal mode? This works normally in my Chrome and Firefox ... Could you give me an image of it?

minaminao replied:
Hmm, but you said it wasn't broken in Firefox.

>It does, at least in Firefox. It overrides color on both a and a:visited. I think you need to style a.remix-link:visited separately.

Are you saying that it doesn't actually break, but it's better to specify `:visited`? I don't know the reason, or I'm not sure I understand what you are saying. If you are saying that `:visited` should be a different style, then [this comment](https://github.com/ethereum/solidity/pull/14122#discussion_r1168060092) is a reply.

minaminao replied:
I've deleted the `!important`s.

I also removed `!important` from the custom.css. It's a small change, so I included it in this PR. There are a lot of `!importants` in dark.css, so I didn't change it in this.

cameel commented on docs/ext/remix_code_links.py:
```diff
@@ -49,22 +42,24 @@ def insert_remix_link(app, doctree, solidity_version):
-
-            insert_node_before(literal_block_node, build_remix_link_node(remix_url))
+        if language not in ['solidity', 'yul']:
+                continue
```
```suggestion
            continue
```

minaminao committed:
fix: remove unnecessary spaces

Co-authored-by: Kamil Śliwak <cameel2@gmail.com>

minaminao committed:
fix: remove unnecessary !important in link

minaminao committed:
refactor: remove !important in custom.css

r0qs reviewed (changes requested):
Overall it looks good to me considering @cameel suggestions.

But I noticed a small problems in the layout. Here, and in most of the other places, the `remix-link-container` class seems to have no margin:

![2023-04-18-141901_4176x1504_scrot](https://user-images.githubusercontent.com/457348/232777134-0c5052fc-64c9-436d-8f47-7be99aed95be.png)

But on the `ir-breaking-changes.html` we have some extra margin in the `remix-link-container` css class:

![2023-04-18-141858_4176x1504_scrot](https://user-images.githubusercontent.com/457348/232777142-7feb229f-1a1d-4ebd-93d4-4529f70b66d1.png)

It would be nice to have this normalized.

minaminao committed:
fix: `.. code-block:: Solidity` -> `.. code-block:: solidity`

minaminao committed:
fix: broken layout of remix button in list items

minaminao commented:
Thank you! It seems that remix-link-container in a `<li>` tag was broken. I fixed it. (I replaced `. code-block:: Solidity` with `. code-block:: solidity` for consistency and use of `.highlight-solidity`.)

minaminao committed:
refactor: make selectors more strict

r0qs reviewed (approved):
Nice! Thanks @minaminao. LGTM :)

JamesonDiamond1 reviewed (commented):
@JamesonDiamond1
